fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/translation/MaxMatchMap.java

```diff
@@ -102,117 +224,789 @@ public ValueEquivalence getValueEquivalence() {
+            }
+        }
+    }
+
+    /**
+     * Class that functions as a partial matcher between a {@link Value} on the query and a reachable {@link Value} on
+     * the candidate side.
+     * <br>
+     * Normally, we establish semantic equality between two {@link Value} by considering not just the values themselves
+     * but the subtrees rooted at these values. In order to compute semantic equality, we call
+     * {@link Value#equalsWithoutChildren(Value)} and if successful recurse into the subtrees and establish semantic
+     * equality for the children. The approach this class takes is somewhat the other way around. For a value
+     * {@link #currentQueryValue} and an initial candidate side root value, this class establishes all potential matches
+     * that are reachable from the candidate root value for this value (sans subtree). These matches may not turn out
+     * to be correct full matches but so far we cannot disprove them to be correct. We then descend into a child on the
+     * query side and try to find all candidate side sub values using the already established potential matches whose
+     * child at the same child slot also potentially matches. This can be repeated. Naturally, the number of potential
+     * matches decreases as we descend a path on the query side that incurs fewer and fewer matching counterparts on the
+     * candidate side. Eventually, we end up at a leaf value and still have potential matches or the list of potential
+     * matches is empty and any further descending will only continue to produce empty match lists.
+     * <br>
+     * Example 1: We attempt to match {@code rcv(rcv(q.x, q.y), rcv(q.a, q.b))} to {@code rcv(q.x, rcv(q.a, q.c))}.
+     * On the top level, the {@code rcv(■, ■)} on the query side potentiall matches the {@code rcv(q.x, rcv(q.a, q.c))}
+     * on the candidate side at root level ({@code ■} denoting an opaque, not-yet considered subtree). Let's say, we now
+     * descend into the second child on the query side. {@code rcv(■, rcv(■, ■))} still partially matches
+     * {@code rcv(■, rcv(q.a, q.c))} on the candidate side. We now descend into the first child of the query side and
+     * establish that there is still a match between: {@code rcv(■, rcv(q.a, ■))} and {@code rcv(■, rcv(q.a, ■))} on the
+     * candidate side. If we had descended into the second child instead, we would have not matched anything, though:
+     * {@code rcv(■, rcv(■, q.b))} does not match {@code rcv(■, rcv(■, q.c))}.
+     * <br>
+     * Example 2: We attempt to match {@code rcv(q.a, q.b))} to {@code rcv(q.x, rcv(q.a, q.b), rcv(q.a, q.c))}.
+     * On top level, nothing matches to the top level of the candidate side. We can, however, establish two
+     * reachable potential matches for {@code rcv(■, ■)} on the candidate side: {@code rcv(q.a, q.b)} and
+     * {@code rcv(q.a, q.b)}. When we now descend further into the second child on the query side, one of these two
```

Comment:
```suggestion
     * {@code rcv(q.a, q.c)}. When we now descend further into the second child on the query side, one of these two
```

Reply:
Just testing if you are already asleep. :-)

Reply:
done.